### ISSUE TITLE
The timing seems one step off?

### ISSUE THREAD

**Issue by Nuxij** [jenkinsci/build-time-blame-plugin#14]
Feel free to call me out, maybe I'm missing something.

Because of the way you match the time taken to the regex lines, the graph reads incorrectly in my opinion.

When a line matches a regex, Time Taken is recorded as the time between the current and previous Elapsed Time. Naturally the line you match is the start of your build step, something like "Calling Unit Tests". This means that the labelled block in the graph is incorrect. It'll be named "Unit Tests" but actually show the Time Taken **until** that line, i.e. the previous step/label.

I hope that makes sense? I work around it by naming the Labels "incorrectly" against the regex, naming them for the step that just finished not what they're finding to start.

The fix is either to always print a "Done" message after every build step and match those instead, or manage the time slightly differently so the graph can read correctly.

**Comment by LTegtmeier** [comment]
This implementation is as intended. It was a better fit for the projects the plugin was initially developed to analyze. Also, the code for this kind of matching is simpler since the code only has to look at the step that actually matched rather than the previous one in the list. 

I agree that it would be easier to understand the Plugin if it worked as you described. Though, it would be a major change to anyone using the Report to change it now. 

From the Readme:
> Regex Search Key: Lines in the console output that match this regex string will be added to the report. These keys should match when the particular step is complete.

**Comment by DKroot** [comment]
I have a very similar need and I'm not sure how to work around it. I was hoping to use multi-line regexes for that, but they don't seem working. I'm trying to profile Postgres `psql` output, which looks like this:

```
0:09:09.858 CREATE TABLE bowen9 AS SELECT a.all_wosids,b.source_id 
0:09:09.858 FROM bowen8 a 
0:09:09.858 INNER JOIN wos_references b ON a.all_wosids=b.cited_source_uid 
0:09:09.858 WHERE b.source_id in (select all_wosids from bowen8);
0:11:58.508 SELECT 381636
```

I can easily match the first line as `bowen9` step. But the time actually spent on this step (almost 3 minutes) gets counted in the next step, not `bowen9`. I can't really change Postgres output unless I explicitly insert `\echo Done: bowen9` and do it for every step. And multi-line regex `(?s)` doesn't seem to work.

An option to "Use labels as start times" would help *a lot*. This could be global or per job or per label. Come to think about it, most logs are easier to match in this mode because script echoing could be turned on for the entire script in most script languages. (e.g. Bash: `set -x`, psql: `\set ECHO all`, etc.). Once echoing is on, it's easy to match source lines without having to do additional "Done: {step}" output.

If this new option is off by default, it'd not break existing reports.

**Comment by LTegtmeier** [comment]
> And multi-line regex (?s) doesn't seem to work.

Yeah, the checking is all one line at a time.

> most logs are easier to match in this mode

I agree. The code for matching is mostly what's simpler with end-based matching.

> If this new option is off by default, it'd not break existing reports.

I'll consider that option. I'd almost prefer a breaking change to force the new behavior. The code for end-based matching wouldn't be able to share much with the new code for start-based matching. Maintaining both branches is bad long term. But, a brief window of code that works both ways could be useful.

**Comment by DKroot** [comment]
I'd be happier as long as I get this option in any form and on any level(s) you think to be reasonable.

**Comment by Nuxij** [comment]
I haven't looked at the code too much but once you've collected the list of matching lines and their timestamps, how simple would it be to optionally run through and just shift the times? Something like:

```python
for match, index in matches:
    match['start'] = match['end']
    match['end'] = matches[index+1]['start']
```

**Comment by LTegtmeier** [comment]
Something like that could work. Dealing with the last match and that the timestamps aren't stored quite that way would require some tweaks. I'm not sure if I like complicating the existing code any further. It's already hard to follow for me and I wrote it. TBH, it's low on my priority list. But, my plan was to essentially rewrite it to clean it up and add this functionality.

**Comment by LTegtmeier** [comment]
🚀 This is released in [v2.0.0](https://github.com/jenkinsci/build-time-blame-plugin/releases/tag/v2.0.0). See those notes and the Readme for more details.
